src/index.ts

```diff
@@ -1,4 +1,6 @@
-export default function alwaysArray<T>(input: T | T[]): T[] {
+export type SingleOrArray<T> = T | readonly T[];
+
+export default function alwaysArray<T>(input: SingleOrArray<T>): readonly T[] {
```

Comment:
What about removing the visibility from the return value?

```suggestion
export default function alwaysArray<T>(input: SingleOrArray<T>): T[] {
```

From my testing, readonly attributes can still be set to non-readonly values:

```ts
interface TestObject {
	readonly value: number[];
}

const a: TestObject = {
	value: [1, 2, 3, 4, 5]
};
const b = alwaysArray(a.value); // [1, 2, 3, 4, 5]
const c: TestObject = {
	value: b // seems to be okay
};
```

Reply:
I think you mixed up the `readonly` keyword here.
The interface in your example has a readonly member with a normal array. It can be set only once but the array itself is writeable.

```ts
interface TestObjectA {
	readonly value: number[];
}

interface TestObjectB {
	value: readonly number[];
}

const a: TestObjectA = {value: [1, 2, 3]};
const b: TestObjectB = {value: [1, 2, 3]};

a.value = []; // fails as the member is readonly
b.value = []; // works as the member is not readonly

a.value.push(5); // works as the variable itself is writeable
b.value.push(5); // is denied as the variable itself is readonly
```

Maybe this example can explain why I requested this change:

```ts
// Method to get some info of an array. This method will not modify the array, so it can be readonly (and should be to let the compiler warn me about changing it.)
function onlyRead(values: readonly number[]): number {
	const foo = someTotallyUnrelatedMethodInATotallyDifferentContext(values)
	const bar = other(values) // does not work
	return foo * bar
}

function someTotallyUnrelatedMethodInATotallyDifferentContext(values: number | readonly number[]) {
	const arr = alwaysArray(values)
	return arr.length
}

function other(values: number | number[]) {
	const arr = alwaysArray(values)
	return arr.length
}
```

Currently the method returns the array directly when the argument is an array: `Array.isArray(input) === true` → `return input`. This returns the input with its type without changing it (readonly array). The return type of the method has to be the same.
The return type could be opened up to be writeable too (`Array.isArray(input) ? […input] : [input]`) but this would introduce an additional content "copy" (as far as JavaScript actually copies and not just references stuff).

As this is used on arguments for methods anyway I don't think it should be writeable anyway so I would keep It readonly.

I hope I could explain my point in a better way now :)

Reply:
Ah, I see. I'd prefer returning a writeable array just to avoid confusion. Maybe we can combine this with #11, where I implemented spreading of iterables, so adding that for arrays as well was trivial: https://github.com/krmax44/always-array/blob/3675f939277f99d721b27d4848ea3c0b4aaac6e6/src/index.ts

Does this solve the issue for you?